helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -73,6 +75,73 @@ public RoutingTableProvider() {
   }
 
   /**
+   * returns the instances for {resource,partition} pair that are in a specific {state} if
+   * aggregateGrouping is turned on, find all resources belongs to the given resourceGroupName and
+   * aggregate all partition states from all these resources.
+   *
+   * @param resourceName
+   * @param partitionName
+   * @param state
+   * @param groupingEnabled
+   *
+   * @return empty list if there is no instance in a given state
+   */
+  public List<InstanceConfig> getInstances(String resourceName, String partitionName, String state,
+      boolean groupingEnabled) {
+    if (!groupingEnabled) {
+      return getInstances(resourceName, partitionName, state);
+    }
+
+    List<InstanceConfig> instanceList = null;
+    RoutingTable _routingTable = _routingTableRef.get();
+    ResourceGroupInfo resourceGroupInfo = _routingTable.getResourceGroup(resourceName);
+    if (resourceGroupInfo != null) {
+      PartitionInfo keyInfo = resourceGroupInfo.get(partitionName);
+      if (keyInfo != null) {
+        instanceList = keyInfo.get(state);
+      }
+    }
+    if (instanceList == null) {
+      instanceList = Collections.emptyList();
+    }
+    return instanceList;
+  }
+
+  /**
+   * returns the instances for {resource,partition} pair that are in a specific {state} if
+   * aggregateGrouping is turned on, find all resources belongs to the given resourceGroupName and
+   * having the given resource tags and return the aggregated partition states from all these
+   * resources.
+   *
+   * @param resourceGroupName
+   * @param partitionName
+   * @param state
+   * @param resourceTags
+   *
+   * @return empty list if there is no instance in a given state
+   */
+  public List<InstanceConfig> getInstances(String resourceGroupName, String partitionName,
```

Comment:
Once the isGroupRoutingEnabled is true in IdealState, it implicitly means:
1) All resource in the same resource group will have same ResourceGroupName
2) Resource tag is identical in resources belongs to same group (resource tag is the way to differentiate different resources in the same resource group).
3) The partition name for all resources in the same group are same (otherwise, they will be treated as different partition in routing table).